[JIT] pickle serialization for custom bound classes

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32658 [JIT] Fix stateful lambda stuff and simplify code in custom C++ binding API
* **#32604 [JIT] pickle serialization for custom bound classes**
* #32586 [JIT] Include custom_class.h in torch/script.h
* #32555 [JIT] Serialize attributes and types in ClassType serialization
* #32553 [JIT] Support for registering C++ lambdas as methods on custom C++ class

Differential Revision: [D19566633](https://our.internmc.facebook.com/intern/diff/D19566633)

Commit:
[JIT] pickle serialization for custom bound classes

[ghstack-poisoned]

Commit:
Update on "[JIT] pickle serialization for custom bound classes"




[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 6075540d:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=6075540d913e14eab02f695e57095c1380df79ea).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4337763) pytorch_linux_xenial_py3_clang5_mobile_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=90570301) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=90570301))
<details>
<summary>
<code>Jan 25 00:34:51 caused by: Connection refused (os error 111)</code>
</summary>

```
Jan 25 00:34:51 ++++ trap -p EXIT 
Jan 25 00:34:51 +++ eval 'extract_trap_cmd ' 
Jan 25 00:34:51 ++++ extract_trap_cmd 
Jan 25 00:34:51 ++++ printf '%s\n' '' 
Jan 25 00:34:51 +++ printf '%s\n' cleanup 
Jan 25 00:34:51 ++ trap -- ' 
Jan 25 00:34:51 cleanup' EXIT 
Jan 25 00:34:51 ++ which sccache 
Jan 25 00:34:51 ++ sccache --stop-server 
Jan 25 00:34:51 error: couldn't connect to server 
Jan 25 00:34:51 caused by: Connection refused (os error 111) 
Jan 25 00:34:51 Stopping sccache server... 
Jan 25 00:34:51 ++ true 
Jan 25 00:34:51 ++ rm /var/lib/jenkins/sccache_error.log 
Jan 25 00:34:51 rm: cannot remove '/var/lib/jenkins/sccache_error.log': No such file or directory 
Jan 25 00:34:51 ++ true 
Jan 25 00:34:51 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Jan 25 00:34:51 ++ SCCACHE_IDLE_TIMEOUT=1200 
Jan 25 00:34:51 ++ RUST_LOG=sccache::server=error 
Jan 25 00:34:51 ++ sccache --start-server 
Jan 25 00:34:51 Starting sccache server... 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Update on "[JIT] pickle serialization for custom bound classes"


Differential Revision: [D19566633](https://our.internmc.facebook.com/intern/diff/D19566633)

[ghstack-poisoned]

Commit:
Update on "[JIT] pickle serialization for custom bound classes"


Differential Revision: [D19566633](https://our.internmc.facebook.com/intern/diff/D19566633)

[ghstack-poisoned]

smessmer commented on c10/util/TypeTraits.h:
```diff
@@ -130,6 +130,25 @@ struct is_type_condition<C, std::enable_if_t<std::is_same<bool, std::remove_cv_t
 template <class T>
 struct is_fundamental : std::is_fundamental<T> {};
 
+/// Strip the class from a method type
```
We already have this functionality, see infer_function_traits in Metaprogramming.h

Commit:
Update on "[JIT] pickle serialization for custom bound classes"


Differential Revision: [D19566633](https://our.internmc.facebook.com/intern/diff/D19566633)

[ghstack-poisoned]

dzhulgakov commented on torch/custom_class.h:
```diff
@@ -121,6 +105,63 @@ class class_ {
+        format_getstate_schema());
+    auto first_arg_type = getstate_schema.arguments().at(0).type();
+    TORCH_CHECK(
+        first_arg_type->isSubtypeOf(classTypePtr),
```
we probably never will get inheritance, but still - can't it be just a comparison check?

dzhulgakov commented on torch/custom_class.h:
```diff
@@ -121,6 +105,63 @@ class class_ {
+    TORCH_INTERNAL_ASSERT(setstate_schema.arguments().size() == 2);
+    auto arg_type = setstate_schema.arguments().at(1).type();
+    TORCH_CHECK(
+        (arg_type->isSubtypeOf(ser_type)),
```
here it should be equality, right?

dzhulgakov commented on torch/custom_class_detail.h:
```diff
@@ -0,0 +1,72 @@
+struct pickle_factory<Get, Set, RetState(Self), NewInstance(ArgState)> {
+  pickle_factory(Get&& g, Set&& s) : g(std::move(g)), s(std::move(s)) {}
+
+  [[noreturn]] ArgState arg_state_type() {}
```
nit: why do you need `decltype(arg_state_type())` instead of just putting here `using arg_state = ArgState` and doing `decltype(pickle_factor).arg_state`?

dzhulgakov reviewed (commented):
Looks good, I'll let Zach to take a look too

Commit:
Update on "[JIT] pickle serialization for custom bound classes"


Differential Revision: [D19566633](https://our.internmc.facebook.com/intern/diff/D19566633)

[ghstack-poisoned]

Commit:
Update on "[JIT] pickle serialization for custom bound classes"


Differential Revision: [D19566633](https://our.internmc.facebook.com/intern/diff/D19566633)

[ghstack-poisoned]

Commit:
Update on "[JIT] pickle serialization for custom bound classes"


Differential Revision: [D19566633](https://our.internmc.facebook.com/intern/diff/D19566633)

[ghstack-poisoned]

Commit:
Update on "[JIT] pickle serialization for custom bound classes"


Differential Revision: [D19566633](https://our.internmc.facebook.com/intern/diff/D19566633)

[ghstack-poisoned]

Commit:
Update on "[JIT] pickle serialization for custom bound classes"


Differential Revision: [D19566633](https://our.internmc.facebook.com/intern/diff/D19566633)

[ghstack-poisoned]

Commit:
Update on "[JIT] pickle serialization for custom bound classes"


Differential Revision: [D19566633](https://our.internmc.facebook.com/intern/diff/D19566633)

[ghstack-poisoned]

Commit:
Update on "[JIT] pickle serialization for custom bound classes"


Differential Revision: [D19566633](https://our.internmc.facebook.com/intern/diff/D19566633)

[ghstack-poisoned]

Commit:
Update on "[JIT] pickle serialization for custom bound classes"


Differential Revision: [D19566633](https://our.internmc.facebook.com/intern/diff/D19566633)

[ghstack-poisoned]

dzhulgakov commented on torch/csrc/jit/pickler.cpp:
```diff
@@ -115,6 +115,17 @@ void Pickler::pushIValueImpl(const IValue& ivalue) {
+    std::stringstream err;
+    err << "Cannot serialize custom bound C++ class";
+    if (memorized_class_types_ && memorized_class_types_->size()) {
+      if (auto qualname = memorized_class_types_->back()->name()) {
```
are you sure that the last one is guaranteed to be current class? If so - I'd add a comment about it

jamesr66a replied:
It's guaranteed to the be current class so long as we only ever see a `Capsule` as an attribute of an enclosing `Object`

zdevito replied:
```
ObjectA {
   ObjectB b;
   Capusle c;
}
```
Capsule will get attributed to ObjectB

jamesr66a replied:
That's only possible if a user can explicitly put a Capsule in a class, which I don't think is possible

dzhulgakov commented on torch/custom_class.h:
```diff
@@ -127,6 +109,63 @@ class class_ {
+        format_getstate_schema());
+    auto ser_type = getstate_schema.returns().at(0).type();
+    auto setstate_schema = classTypePtr->getMethod("__setstate__")->getSchema();
+    TORCH_INTERNAL_ASSERT(setstate_schema.arguments().size() == 2);
```
shouldn't it be TORCH_CHECK too?

jamesr66a replied:
This is the number of arguments for the wrapper we generated above, not the user's code, so it wouldn't be meaningful to TORCH_CHECK on that

dzhulgakov commented on torch/custom_class_detail.h:
```diff
@@ -0,0 +1,70 @@
+      c10::guts::is_stateless_lambda<std::decay_t<GetState>>::value &&
+          c10::guts::is_stateless_lambda<std::decay_t<SetState>>::value,
+      "torch::jit::pickle_ currently only supports lambdas as "
+      "__getstate__ and __setstate__ arguments.");
```
just curious - what prevents us from passing raw functions? aren't we wrapping the whole thing into std::function anyway?

jamesr66a replied:
Pybind seems to only support lambdas as well ;)

This restriction was from me trying to not go down the rabbithole of supporting (lambda, lambda), (lambda, function), (function, lambda), (function, function) in template metaprogramming land. maybe someone who wants to think about that can implement it

dzhulgakov commented on torch/custom_class_detail.h:
```diff
@@ -0,0 +1,70 @@
+template <
+    typename Get,
+    typename Set,
+    typename = typename c10::guts::infer_function_traits<Get>::type::func_type,
```
this is magic :)

zdevito commented on torch/custom_class.h:
```diff
@@ -127,6 +109,63 @@ class class_ {
+    TORCH_CHECK(
+        *first_arg_type == *classTypePtr,
+        "self argument of __getstate__ must be the custom class type. Got ",
+        first_arg_type->str());
```
python_str

zdevito commented on torch/custom_class.h:
```diff
@@ -127,6 +109,63 @@ class class_ {
+        (*arg_type == *ser_type),
+        "__setstate__'s argument should be the same type as the "
+        "return value of __getstate__. Got ",
+        arg_type->str(),
```
python_str

zdevito commented on torch/custom_class.h:
```diff
@@ -127,6 +109,63 @@ class class_ {
+        "return value of __getstate__. Got ",
+        arg_type->str(),
+        " but expected ",
+        ser_type->str());
```
python_str

zdevito commented on test/cpp/jit/test_custom_class.cpp:
```diff
@@ -83,8 +75,14 @@ static auto testStack =
         .def("merge", &Stack<std::string>::merge)
-        .def("__getstate__", &Stack<std::string>::__getstate__)
-        .def("__setstate__", &Stack<std::string>::__setstate__)
+        .def(torch::jit::pickle_(
```
people will confuse torch::jit::pickle with torch::jit::pickle_. Compromise would be to make it `.def_pickle(get, set);` which is close enough to pybind but clearly different from the pickler.

zdevito reviewed (approved):
Some minor comments but generally looks good.

Commit:
Update on "[JIT] pickle serialization for custom bound classes"


Differential Revision: [D19566633](https://our.internmc.facebook.com/intern/diff/D19566633)

[ghstack-poisoned]

Commit:
Update on "[JIT] pickle serialization for custom bound classes"


Differential Revision: [D19566633](https://our.internmc.facebook.com/intern/diff/D19566633)

[ghstack-poisoned]
